src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -354,7 +354,6 @@ class NBDServer
     void* entry() override
     {
       (server.*func)();
-      server.shutdown();
```

Comment:
Yes, shutdown is called in "stop" but it does not reach there.

After `ioctl(nbd, NBD_DISCONNECT)` the reader thread gets NBD_CMD_DISC and just returns. But the main thread (do_map function) is still waiting on `ioctl(nbd, NBD_DO_IT)`. For this ioctl to terminate, one need to shutdown the socket. "stop" is run after `ioctl(nbd, NBD_DO_IT)`.

I can't see how it may work for you.
Also I think `shutdown` in stop can be removed, but this (after entry function return) should remain. Previosly I suggested to add `shutdown()` in reader_entry after "disconnect request received" instead, but it is not correct. It will work only for "normal" exit. In the case of error return we will not call shutdown and will hang.